### PR TITLE
Convert functions to accept RequestHandlers

### PR DESCRIPTION
**Pull request by jaidevd**
See #304 
This PR adds a decorator: `gramex.handlers.functionhandler.add_handler` which wraps any callable to:

1. get it's args and kwargs from handler.path_args and handler.request.body
2. return a JSON object
3. enforce any type annotations found in the callable

ToDo:
- [x] ~~Not exhaustively tested for POST / DELETE~~

### CODE REVIEW

File: gramex/handlers/functionhandler.py [hunk 1]
```diff
@@ -1,13 +1,49 @@
...
+    func : callable
+        function to be wrapped.
+    """
+    sig = signature(func)
+
+    @wraps(func)
+    def wrapper(handler):
+        args = handler.path_args
+        if handler.request.method == 'GET':
+            kwargs = {k: v[0] for k, v in handler.args.items()}
```
Comment (jaidevd) [inline gramex/handlers/functionhandler.py:31]: `handler.get_arg` does this.

Comment (jaidevd) [reply gramex/handlers/functionhandler.py:31]: Test all HTTP methods.
Test all native types including `None`.
Test what happens with default named arguments.
`?x=1&x=2&x=3` should be doable. Perform the typecasting from the hints.

Comment (pratapvardhan) [inline gramex/handlers/functionhandler.py:31]: Perhaps `{k: v if len(v) > 1 else v[0] for k, v in handler.args.items()}` to handle `?x=1&x=2&y=3`?

File: gramex/handlers/functionhandler.py [hunk 2]
```diff
@@ -1,13 +1,49 @@
...
+            kwargs = handler.body
+        hints = get_type_hints(func)
+        for arg, argtype in hints.items():
+            param = sig.parameters.get(arg, False)
+            if param:
+                if param.kind == param.VAR_POSITIONAL:
+                    args = [argtype(k) for k in args]
+                else:
+                    kwargs[arg] = argtype(kwargs[arg])
+        return json.dumps(func(*args, **kwargs))
```
Comment (pratapvardhan) [inline gramex/handlers/functionhandler.py:42]: Thoughts on using pydantic? See this example https://pydantic-docs.helpmanual.io/usage/validation_decorator/

File: gramex/handlers/functionhandler.py (hunk 1 above)
Comment (jaidevd) [reply gramex/handlers/functionhandler.py:31]: @sanand0 This ^ works. `handler.get_arg('x')` only gets the last one.

**Comment by jaidevd** [comment]
@sanand0 items from the last review are done. Some notes on usage are in the [docstring for `add_handler`](https://github.com/gramener/gramex/blob/5c57e1106e41d93a2583d728757502df33d7a0c9/gramex/handlers/functionhandler.py#L41). Please take a look.

File: gramex/handlers/functionhandler.py (hunk 2 above)
Comment (jaidevd) [reply gramex/handlers/functionhandler.py:125]: @pratapvardhan this is great... I can offload all the typechecking / casting off to `pydantic.validate_arguments`.
Type validation should be on by default, and in that case, pydantic should do it. But users should also be able to turn it off if required, like:
```python
@add_handler(validate=False)
def slow_sum(*numbers: float, delay: int):
   s = 0
   for i in numbers:
       time.sleep(delay)
       s += 0
       yield s
```
This ^ snippet won't complain when `numbers` is not floats, say.

@sanand0 Thoughts?

Comment (sanand0) [reply gramex/handlers/functionhandler.py:125]: No inputs -- and we can defer this.

**Comment by sanand0** [comment]
- [ ] Move `add_handler` as `gramex.transforms.handler`
- [ ] Fix typos in docstring
    - `greet.,py` -> `greet.py`
    - `funtion` -> `function`
    - `andd` -> `and`
    - Please check for others
- [ ] Change usage to:

```yaml
pattern: /$YAMLURL/greet
handler: Function
kwargs:
    function: greet.birthday
```

- [ ] This test case should

```python
def total_list(items: List[float], start: float) -> float:
    return sum(items) + start
```

- [ ] In POST and PUT requests, URL parameters and path arguments should be used. Add test cases

**Comment by sanand0** [comment]
These scenarios don't yet work.

```yaml
url:
  power:
    pattern: /power
    handler: FunctionHandler
    kwargs:
      function: alpha.power
      kwargs:
        x: 3

  message:
    pattern: /message
    handler: FunctionHandler
    kwargs:
      function: alpha.messages
```

```python
@handler
def power(x: int, y:float) -> float:
    return y ** x


@handler
def messages(s: list):
    for msg in s:
        yield msg
```

**Comment by jaidevd** [comment]
@sanand0 Added these two cases, and tests for them.

**Comment by sanand0** [comment]
I made these changes:

- [x] Make all handler names, URL patterns and function names identical
- [x] Move `@handler` methods from `test_function.py` to `functionutils.py`. This is to allow manual testing. Specifically, if `cd tests` and run `gramex` to visit `localhost:9999`, Gramex can't import `tests.test_function` because it has a line `from . import TestGramex`.
- [x] Split tests into a `test_functionhandler.TestWrapper` class so we can run them independently
- [x] If `?x=1&x=2` are passed to `func(x: int)`, then the **last** parameter is used, not the first
- [x] POST body is processed as JSON only if `Content-Type: application/json`. This is consistent with FormHandler. It also retains the URL query parameters apart from the body
- [x] If we call `/func/total/10/20?items=30&items=40` it will add 10+20+30+40=100. THis required a rewrite

**Comment by sanand0** [comment]
Note: Checks failed because Travis couldn't access the Microsoft R repository. Even [master fails](https://travis-ci.com/github/gramener/gramex/builds/191618600) despite the same commit passing earlier. So I've gone ahead and merged.

**Pull request merged**